include/ringmesh/mesh/mesh_base.h

```diff
@@ -0,0 +1,136 @@
+ *     54518 VANDOEUVRE-LES-NANCY
+ *     FRANCE
+ */
+
+#pragma once
+
+#include <ringmesh/mesh/common.h>
+
+#include <algorithm>
+#include <memory>
+
+#include <ringmesh/basic/factory.h>
+#include <ringmesh/basic/nn_search.h>
+
+#include <ringmesh/mesh/mesh_aabb.h>
+
+namespace GEO
+{
+    class AttributesManager;
+} // namespace GEO
+
+namespace RINGMesh
+{
+    FORWARD_DECLARATION_DIMENSION_CLASS( GeoModel );
+    FORWARD_DECLARATION_DIMENSION_CLASS( MeshBaseBuilder );
+    FORWARD_DECLARATION_DIMENSION_CLASS( PointSetMeshBuilder );
+    FORWARD_DECLARATION_DIMENSION_CLASS( LineMeshBuilder );
+    FORWARD_DECLARATION_DIMENSION_CLASS( SurfaceMeshBuilder );
+    FORWARD_DECLARATION_DIMENSION_CLASS( VolumeMeshBuilder );
+    FORWARD_DECLARATION_DIMENSION_CLASS( SurfaceMesh );
+
+    struct EdgeLocalVertex;
+    struct ElementLocalVertex;
+    struct PolygonLocalEdge;
+    struct CellLocalFacet;
+} // namespace RINGMesh
+
+namespace RINGMesh
+{
+    /*!
+     * class base class for encapsulating Mesh structure
+     * @brief encapsulate adimensional mesh functionalities in order to provide
+     * an API
+     * on which we base the RINGMesh algorithms
+     * @note For now, we encapsulate the GEO::Mesh class.
+     */
+    // BEGINING OF MESHBASE DEFINITION
+    template < index_t DIMENSION >
+    class MeshBase
+    {
+        ringmesh_disable_copy_and_move( MeshBase );
+        ringmesh_template_assert_2d_or_3d( DIMENSION );
+        friend class MeshBaseBuilder< DIMENSION >;
+
+    public:
+        virtual ~MeshBase() = default;
+
+        virtual void save_mesh( const std::string& filename ) const = 0;
+
+        virtual std::tuple< index_t, std::vector< index_t > >
+            connected_components() const = 0;
+
+        /*!
+         * \name Vertex methods
+         * @{
+         */
+        /*!
+         * @brief Gets a point.
+         * @param[in] v_id the vertex, in 0.. @function nb_vetices()-1.
+         * @return const reference to the point that corresponds to the vertex.
+         */
+        virtual const vecn< DIMENSION >& vertex( index_t v_id ) const = 0;
+        /*
+         * @brief Gets the number of vertices in the Mesh.
+         */
+        virtual index_t nb_vertices() const = 0;
+
+        virtual GEO::AttributesManager& vertex_attribute_manager() const = 0;
+
+        /*!
+         * @brief return the NNSearch at vertices
+         * @warning the NNSearch is destroyed when calling the
+         * Mesh::polygons_aabb()
+         * and Mesh::cells_aabb()
+         */
+        const NNSearch< DIMENSION >& vertex_nn_search() const;
+
+        virtual MeshType type_name() const = 0;
+
+        virtual std::string default_extension() const = 0;
+
+        virtual bool is_mesh_valid() const = 0;
+
+        /*!
+         * @}
+         */
+    protected:
+        MeshBase() = default;
+
+    protected:
+        mutable std::unique_ptr< NNSearch< DIMENSION > > vertex_nn_search_{};

```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Make this member variable "private". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=cpp%3AS3656)